plugins/inputs/tail/tail.go

```diff
@@ -290,25 +310,62 @@ func (t *Tail) tailNewFiles() error {
 			// create a goroutine for each "tailer"
 			t.wg.Add(1)
 
-			go func() {
+			// Store the tailer in the map before starting the goroutine
+			t.tailersMutex.Lock()
+			t.tailers[tailer.Filename] = tailer
+			t.tailersMutex.Unlock()
+
+			go func(tl *tail.Tail) {
 				defer t.wg.Done()
-				t.receiver(parser, tailer)
+				t.receiver(parser, tl)
 
-				t.Log.Debugf("Tail removed for %q", tailer.Filename)
+				t.Log.Debugf("Tail removed for %q", tl.Filename)
 
-				if err := tailer.Err(); err != nil {
+				if err := tl.Err(); err != nil {
 					if strings.HasSuffix(err.Error(), "permission denied") {
-						t.Log.Errorf("Deleting tailer for %q due to: %v", tailer.Filename, err)
-						delete(t.tailers, tailer.Filename)
+						t.Log.Errorf("Deleting tailer for %q due to: %v", tl.Filename, err)
+						t.tailersMutex.Lock()
+						delete(t.tailers, tl.Filename)
+						t.tailersMutex.Unlock()
 					} else {
-						t.Log.Errorf("Tailing %q: %s", tailer.Filename, err.Error())
+						t.Log.Errorf("Tailing %q: %s", tl.Filename, err.Error())
 					}
 				}
-			}()
+			}(tailer)
+		}
+	}
 
-			t.tailers[tailer.Filename] = tailer
+	// Clean up tailers for files that are no longer being monitored
+	t.tailersMutex.Lock()
+	for file, tailer := range t.tailers {
+		if !currentFiles[file] {
+			// This file is no longer in our glob pattern matches
+			// We need to stop tailing it and remove it from our list
+			t.Log.Debugf("Removing tailer for %q as it's no longer in the glob pattern", file)
+
+			// Save the current offset for potential future use
+			if !t.Pipe {
+				offset, err := tailer.Tell()
+				if err == nil {
+					t.Log.Debugf("Recording offset %d for %q", offset, tailer.Filename)
+					t.offsets[tailer.Filename] = offset
+				} else {
+					t.Log.Errorf("Recording offset for %q: %s", tailer.Filename, err.Error())
+				}
+			}
```

Comment:
Shouldn't this be done when reading the file? Otherwise the offset might be lost if Telegraf goes down without stop tailing a file...